ddht/rpc.py

```diff
@@ -0,0 +1,253 @@
+from abc import abstractmethod
+import collections
+import io
+import json
+import logging
+import pathlib
+from typing import Any, Dict, Generic, Mapping, Optional, Tuple, TypeVar, cast
+
+from async_service import Service
+from eth_utils import ValidationError
+import trio
+
+from ddht.abc import RPCHandlerAPI, RPCRequest, RPCResponse
+from ddht.typing import JSON
+
+NEW_LINE = "\n"
+
+
+def strip_non_json_prefix(raw_request: str) -> Tuple[str, str]:
+    if raw_request and raw_request[0] != "{":
+        prefix, bracket, rest = raw_request.partition("{")
+        return prefix.strip(), bracket + rest
+    else:
+        return "", raw_request
+
+
+logger = logging.getLogger("ddht.rpc")
+decoder = json.JSONDecoder()
+
+
+async def read_json(
+    socket: trio.socket.SocketType,
+    buffer: io.StringIO,
+    decoder: json.JSONDecoder = decoder,
+) -> JSON:
+    request: JSON
+
+    while True:
+        data = await socket.recv(1024)
+        buffer.write(data.decode())
+
+        bad_prefix, raw_request = strip_non_json_prefix(buffer.getvalue())
+        if bad_prefix:
+            logger.info("Client started request with non json data: %r", bad_prefix)
+            await write_error(socket, f"Cannot parse json: {bad_prefix}")
+            continue
+
+        try:
+            request, offset = decoder.raw_decode(raw_request)
+        except json.JSONDecodeError:
+            # invalid json request, keep reading data until a valid json is formed
+            if raw_request:
+                logger.debug(
+                    "Invalid JSON, waiting for rest of message: %r", raw_request,
+                )
+            else:
+                await trio.sleep(0.01)
+            continue
+
+        # TODO: more efficient algorithm can be used here by
+        # manipulating the buffer such that we can seek back to the
+        # correct position for *new* data to come in.
+        buffer.seek(0)
+        buffer.write(raw_request[offset:])
+        buffer.truncate()
+
+        break
+
+    return request
+
+
+async def write_error(socket: trio.socket.SocketType, message: str) -> None:
+    json_error = json.dumps({"error": message})
+    await socket.send(json_error.encode("utf8"))
+
+
+def validate_request(request: Mapping[Any, Any]) -> None:
+    try:
+        version = request["jsonrpc"]
+    except KeyError as err:
+        raise ValidationError("Missing 'jsonrpc' key") from err
+    else:
+        if version != "2.0":
+            raise ValidationError(f"Invalid version: {version}")
+
+    if "method" not in request:
+        raise ValidationError("Missing 'method' key")
+    if "params" in request:
+        if not isinstance(request["params"], list):
+            raise ValidationError("Missing 'method' key")
+
+
+def generate_response(
```

Comment:
Instead of a single method which uses `TypeError` to manually implement polymorphism, this would better lean on the language if it was two separate methods, `generate_success_response` and `generate_error_response`

Reply:
Oh, `RPCHandler` does this :sweat_smile: having the helpers up there means you wouldn't have to re-implement them below, `RPCHandler` would be a little cleaner.